zkvm/src/txlog.rs

```diff
@@ -76,6 +84,73 @@ impl TxID {
         entry.commit_to_transcript(&mut t);
         t.challenge_bytes(b"merkle.leaf", result);
     }
+
+    /// Computes the Merkle proof of inclusion required to validate an
+    /// entry at index i, not including the original item i.
+    /// Fails when requested index is out of bounds
+    pub fn proof(list: &[Entry], index: usize) -> Result<Vec<MerkleHash>, VMError> {
+        if index >= list.len() {
+            return Err(VMError::InvalidMerkleProof);
+        }
+        let t = Transcript::new(b"ZkVM.txid");
+        let mut result = Vec::new();
+        Self::subproof(t, list, index, &mut result)?;
+        Ok(result)
+    }
+
+    fn subproof(
+        t: Transcript,
+        list: &[Entry],
+        index: usize,
+        result: &mut Vec<MerkleHash>,
+    ) -> Result<(), VMError> {
+        match list.len() {
+            0 => Err(VMError::InvalidMerkleProof),
+            1 => Ok(()),
+            n => {
+                let k = n.next_power_of_two() / 2;
+                if index >= k {
+                    let mut lefthash = [0u8; 32];
```

Comment:
maybe i missed something, but seems like `lefthash`/`righthash` remain zeroed and not actually computed?

Reply:
they get set when we pass into `Self::node(...)` !

Reply:
Seems like we are hashing everything quadratic number of times? `::node` recurses by itself, and then we descend one level down, and call `::node` again, that walks over the same hashes.

WDYT of simply building the whole tree and sampling it? Cf. https://github.com/interstellar/slingshot/pull/188/files#r262289600

Reply:
that makes a lot of sense to me! will start working on that :)